File: chaincode/smartLandRegistro/node/smartLandRegistro.js
/*
# Copyright IBM Corp. All Rights Reserved.
#
# SPDX-License-Identifier: Apache-2.0
*/

'use strict';
const shim = require('fabric-shim');
const util = require('util');

let Chaincode = class {

  // The Init method is called when the Smart Contract 'fabcar' is instantiated by the blockchain network
  // Best practice is to have any Ledger initialization in separate function -- see initLedger()
  async Init(stub) {
    console.info('=========== Instantiated smartLandRegistro chaincode ===========');
    return shim.success();
  }

  // The Invoke method is called as a result of an application request to run the Smart Contract
  // 'fabcar'. The calling application program has also specified the particular smart contract
  // function to be called, with arguments
  async Invoke(stub) {
    let ret = stub.getFunctionAndParameters();
    console.info(ret);

    let method = this[ret.fcn];
    if (!method) {
      console.error('no function of name:' + ret.fcn + ' found');
      throw new Error('Received unknown function ' + ret.fcn + ' invocation');
    }
    try {
      let payload = await method(stub, ret.params);
      return shim.success(payload);
    } catch (err) {
      console.log(err);
      return shim.error(err);
    }
  }

  async queryLandOwner(stub, args) {
    if (args.length != 1) {
      throw new Error('Incorrect number of arguments. Expecting OwnerName ex: Rajeev');
    }
    let landOwner = args[0];

    let carAsBytes = await stub.getState(landOwner); //get the car from chaincode state
    if (!carAsBytes || carAsBytes.toString().length <= 0) {
      throw new Error(landOwner + ' does not exist: ');
    }
    console.log(carAsBytes.toString());
    return carAsBytes;
  }

  async initLedger(stub, args) {
    console.info('============= START : Initialize Ledger ===========');
    let land_properties = [];
    land_properties.push({
      land: 'Gajanan Enclave'
      owner: 'Rajeev'
    });
    land_properties.push({
	    land: 'Gautami Enclave'
      owner: 'Mahendra'
    });
    land_properties.push({
      land: 'Celestia'
      owner: 'Sanjeev'
    });
    land_properties.push({
      land: 'SNiwas'
      owner: 'Mahendra'
    });
    land_properties.push({
      land: 'MantriC'
      owner: 'Biman'
    });
    land_properties.push({
      land: 'MantriD'
      owner: 'SmitaT'
    });
    land_properties.push({
      land: 'Mantri Celestia'
      owner: 'Saraswat'
    });
    land_properties.push({
      land: 'VishnuHomes'
      owner: 'Rajeev'
    });
    land_properties.push({
      land: 'SriRamaResdiency'
      owner: 'Rajesh'
    });
    land_properties.push({
      land: 'ChoudharyNiwas'
      owner: 'Choudhary ram'
    });

    for (let i = 0; i < land_properties.length; i++) {
      land_properties[i].docType = 'land';
      await stub.putState('LAND' + i, Buffer.from(JSON.stringify(land_properties[i])));
      console.info('Added <--> ', land_properties[i]);
    }
    console.info('============= END : Initialize Ledger ===========');
  }

  async createLandOwnerRecord(stub, args) {
    console.info('============= START : Create land-owner record ===========');
    if (args.length != 2) {
      throw new Error('Incorrect number of arguments. Expecting 2');
    }

    var landOwnerRec = {
      docType: 'land',
      land: args[0],
      owner: args[1]
    };

    await stub.putState(args[0], Buffer.from(JSON.stringify(landOwnerRec)));
    console.info('============= END : Create land-owner ===========');
  }

  async queryAllland_properties(stub, args) {

    let startKey = 'Celestia';
    let endKey = 'ChoudharyNiwas';

    let iterator = await stub.getStateByRange(startKey, endKey);

    let allResults = [];
    while (true) {
      let res = await iterator.next();

      if (res.value && res.value.value.toString()) {
        let jsonRes = {};
        console.log(res.value.value.toString('utf8'));

        jsonRes.Key = res.value.key;
        try {
          jsonRes.Record = JSON.parse(res.value.value.toString('utf8'));
        } catch (err) {
          console.log(err);
          jsonRes.Record = res.value.value.toString('utf8');
        }
        allResults.push(jsonRes);
      }
      if (res.done) {
        console.log('end of data');
        await iterator.close();
        console.info(allResults);
        return Buffer.from(JSON.stringify(allResults));
      }
    }
  }

  async changeLandOwner(stub, args) {
    console.info('============= START : changeLandOwner ===========');
    if (args.length != 2) {
      throw new Error('Incorrect number of arguments. Expecting 2');
    }

    let carAsBytes = await stub.getState(args[0]);
    let car = JSON.parse(carAsBytes);
    car.owner = args[1];

    await stub.putState(args[0], Buffer.from(JSON.stringify(car)));
    console.info('============= END : changeLandOwner ===========');
  }
};

shim.start(new Chaincode());

File: chaincode/smartLandRegistro/go/smartLandRegistro.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*
 * The sample smart contract for documentation topic:
 * Writing Your First Blockchain Application
 */

package main

/* Imports
 * 4 utility libraries for formatting, handling bytes, reading and writing JSON, and string manipulation
 * 2 specific Hyperledger Fabric specific libraries for Smart Contracts
 */
import (
	"bytes"
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/hyperledger/fabric/core/chaincode/shim"
	sc "github.com/hyperledger/fabric/protos/peer"
)

// Define the Smart Contract structure
type SmartContract struct {
}

// Define the land-owner structure, with 2 properties.  Structure tags are used by encoding/json library
type Land struct {
	Land   string `json:"land"`
	Owner  string `json:"owner"`
}

/*
 * The Init method is called when the Smart Contract "smartLandOwner" is instantiated by the blockchain network
 * Best practice is to have any Ledger initialization in separate function -- see initLedger()
 */
func (s *SmartContract) Init(APIstub shim.ChaincodeStubInterface) sc.Response {
	return shim.Success(nil)
}

/*
 * The Invoke method is called as a result of an application request to run the Smart Contract "smartLandOwner"
 * The calling application program has also specified the particular smart contract function to be called, with arguments
 */
func (s *SmartContract) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {

	// Retrieve the requested Smart Contract function and arguments
	function, args := APIstub.GetFunctionAndParameters()
	// Route to the appropriate handler function to interact with the ledger appropriately
	if function == "queryLand" {
		return s.queryLand(APIstub, args)
	} else if function == "initLedger" {
		return s.initLedger(APIstub)
	} else if function == "createLandOwner" {
		return s.createLandOwner(APIstub, args)
	} else if function == "queryAllLands" {
		return s.queryAllLands(APIstub)
	} else if function == "changeLandOwner" {
		return s.changeLandOwner(APIstub, args)
	}

	return shim.Error("Invalid Smart Contract function name from smaartLandRegistro")
}

func (s *SmartContract) queryLand(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments. Expecting 1")
	}

	carAsBytes, _ := APIstub.GetState(args[0])
	return shim.Success(carAsBytes)
}

func (s *SmartContract) initLedger(APIstub shim.ChaincodeStubInterface) sc.Response {
	lands := []Land{
		Land{Land: "VishnuHomes", Owner: "Rajeev"},
		Land{Land: "SriRamaResidency", Owner: "Mahendra"},
		Land{Land: "GajananEnclave", Owner: "Nitin"},
		Land{Land: "ChirecHome", Owner: "Smita"},
	}

	i := 0
	for i < len(lands) {
		fmt.Println("i is ", i)
		landAsBytes, _ := json.Marshal(lands[i])
		APIstub.PutState("LAND"+strconv.Itoa(i), carAsBytes)
		fmt.Println("Added", lands[i])
		i = i + 1
	}

	return shim.Success(nil)
}

func (s *SmartContract) createCar(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 5 {
		return shim.Error("Incorrect number of arguments. Expecting 2")
	}

	var land = Land{Land: args[1], Owner: args[2]}

	landAsBytes, _ := json.Marshal(land)
	APIstub.PutState(args[0], landAsBytes)

	return shim.Success(nil)
}

func (s *SmartContract) queryAllLands(APIstub shim.ChaincodeStubInterface) sc.Response {

	startKey := "VishnuHomes"
	endKey := "ChirecHome"

	resultsIterator, err := APIstub.GetStateByRange(startKey, endKey)
	if err != nil {
		return shim.Error(err.Error())
	}
	defer resultsIterator.Close()

	// buffer is a JSON array containing QueryResults
	var buffer bytes.Buffer
	buffer.WriteString("[")

	bArrayMemberAlreadyWritten := false
	for resultsIterator.HasNext() {
		queryResponse, err := resultsIterator.Next()
		if err != nil {
			return shim.Error(err.Error())
		}
		// Add a comma before array members, suppress it for the first array member
		if bArrayMemberAlreadyWritten == true {
			buffer.WriteString(",")
		}
		buffer.WriteString("{\"Key\":")
		buffer.WriteString("\"")
		buffer.WriteString(queryResponse.Key)
		buffer.WriteString("\"")

		buffer.WriteString(", \"Record\":")
		// Record is a JSON object, so we write as-is
		buffer.WriteString(string(queryResponse.Value))
		buffer.WriteString("}")
		bArrayMemberAlreadyWritten = true
	}
	buffer.WriteString("]")

	fmt.Printf("- queryAllLands:\n%s\n", buffer.String())

	return shim.Success(buffer.Bytes())
}

func (s *SmartContract) changeLandOwner(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 2 {
		return shim.Error("Incorrect number of arguments. Expecting 2")
	}

	landAsBytes, _ := APIstub.GetState(args[0])
	land := Land{}

	json.Unmarshal(landAsBytes, &car)
	land.Owner = args[1]

	landAsBytes, _ = json.Marshal(car)
	APIstub.PutState(args[0], landAsBytes)

	return shim.Success(nil)
}

// The main function is only relevant in unit test mode. Only included here for completeness.
func main() {

	// Create a new Smart Contract
	err := shim.Start(new(SmartContract))
	if err != nil {
		fmt.Printf("Error creating new Smart Contract: %s", err)
	}
}
